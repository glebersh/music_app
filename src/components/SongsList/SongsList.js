import React from 'react';
import { useSelector } from 'react-redux';
import SongCard from '../SongCard/SongCard';
import { Alert, AlertTitle, AlertDescription, AlertIcon, Spinner } from '@chakra-ui/react';

const SongsList = ({ type }) => {

  const content = useSelector(state => {
    if (type === 'ARTIST_TOP_TRACKS') {
      return state.artistInfoReducer?.artistTopTracks?.tracks;
    }
    return state.searchReducer?.searchResult?.tracks?.items;
  });

  const loadingStatus = useSelector(state => {
    if (type === 'ARTIST_TOP_TRACKS') {
      return state.artistInfoReducer.loadingStatus;
    }
    return state.searchReducer.loadingStatus;
  });

  const errorStatus = useSelector(state => {
    if (type === 'ARTIST_TOP_TRACKS') {
      return state.artistInfoReducer.errorStatus;
    }
    return state.searchReducer.errorStatus;
  });

  return (
    <>
      {errorStatus &&
        <Alert status='error' role='error-alert'>
          <AlertIcon />
          <AlertTitle>Something went wrong...</AlertTitle>
          <AlertDescription>Check the API response</AlertDescription>
        </Alert>}

      {loadingStatus === 'loading' && <Spinner w='150px' h='150px' role='spinner' />}

      {loadingStatus === "resolved" && content.map((item, index) => {
        return (<SongCard
          key={item.id}
          {...item}
          arrayPosition={index}
        />)
      })}
    </>
  )
};

export default SongsList;