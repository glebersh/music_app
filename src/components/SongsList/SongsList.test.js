import SongsList from "./SongsList";
import { render } from "@testing-library/react";
import '@testing-library/jest-dom';
import userEvent from "@testing-library/user-event";

import * as reduxHooks from 'react-redux';
import { BrowserRouter } from "react-router-dom";

jest.mock('react-redux');
const mockedUseSelector = jest.spyOn(reduxHooks, 'useSelector');
const mockedUseDispatch = jest.spyOn(reduxHooks, 'useDispatch');

const mockedData = [
  {
    album: {
      album_type: 'album',
      artists: [
        {
          id: '1z4g3DjTBBZKhvAroFlhOM',
          name: 'Red Velvet',
        }
      ],
      id: '3rVtm00UfbuzWOewdm4iYM',
      images: [
        {
          height: 64,
          url: 'https://i.scdn.co/image/ab67616d00004851df5022bdf1ac4bf52135c4be',
          width: 64
        }
      ],
      name: '‘The ReVe Festival’ Finale',
    },
    artists: [
      {
        external_urls: {
          spotify: 'https://open.spotify.com/artist/1z4g3DjTBBZKhvAroFlhOM'
        },
        id: '1z4g3DjTBBZKhvAroFlhOM',
        name: 'Red Velvet',
      }
    ],
    duration_ms: 210560,
    explicit: false,
    href: 'https://api.spotify.com/v1/tracks/3CYH422oy1cZNoo0GTG1TK',
    id: '3CYH422oy1cZNoo0GTG1TK',
    name: 'Psycho',
    preview_url: 'https://p.scdn.co/mp3-preview/786146ba24ff84e70da6021c141b1ddd843a6fc0?cid=ea5fd0d4cfe147699b87b05db7e6f7bb',
  },
  {
    album: {
      album_type: 'album',
      artists: [
        {
          id: '1z4g3DjTBBZKhvAroFlhOM',
          name: 'Red Velvet',
        }
      ],
      id: '6MNlcai3skKLKv5syzFwC3',
      images: [
        {
          height: 64,
          url: 'https://i.scdn.co/image/ab67616d000048513f30a062dafcdbc1a8fad842',
          width: 64
        }
      ],
      name: 'Russian Roulette - The 3rd Mini Album',
    },
    artists: [
      {
        id: '1z4g3DjTBBZKhvAroFlhOM',
        name: 'Red Velvet',
      }
    ],
    duration_ms: 211243,
    explicit: false,
    id: '5HiSc2ZCGn8L3cH3qSwzBT',
    name: '러시안 룰렛 Russian Roulette',
    preview_url: 'https://p.scdn.co/mp3-preview/4bcebb26f081af2193258ec4dcef59a97a8dad34?cid=ea5fd0d4cfe147699b87b05db7e6f7bb',
  }];

describe('Song list tests', () => {
  it('Should render and match snapshot if loading is \'resovled\'', () => {
    mockedUseSelector
      .mockReturnValueOnce(mockedData)
      .mockReturnValueOnce('resolved')
      .mockReturnValueOnce(null);

    const songsList = render(
      <BrowserRouter>
        <SongsList />
      </BrowserRouter>);
    expect(songsList).toMatchSnapshot();
  });

  it('If loading is \'loading\' should render spinner', () => {
    mockedUseSelector
      .mockReturnValueOnce(mockedData)
      .mockReturnValueOnce('loading')
      .mockReturnValueOnce(null);

    const songsList = render(
      <BrowserRouter>
        <SongsList />
      </BrowserRouter>);
    expect(songsList.getByRole('spinner')).toBeInTheDocument();
  });

  it('Should render alert if error', () => {
    mockedUseSelector
      .mockReturnValueOnce(mockedData)
      .mockReturnValueOnce('resolved')
      .mockReturnValueOnce(true);

    const songsList = render(
      <BrowserRouter>
        <SongsList />
      </BrowserRouter>);
    expect(songsList.getByRole('error-alert')).toBeInTheDocument();
  });
});