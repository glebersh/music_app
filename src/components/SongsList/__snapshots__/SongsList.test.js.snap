// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Song list tests Should render and match snapshot if loading is 'resovled' 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="song-container"
      >
        <div
          class="css-16l185a"
        >
          <div
            class="css-k008qs"
          >
            <img
              class="chakra-image css-1oaqw3l"
              loading="lazy"
              src="https://i.scdn.co/image/ab67616d00004851df5022bdf1ac4bf52135c4be"
            />
            <div
              class="css-13yz2if"
            >
              <div
                class="css-0"
              >
                <p
                  class="chakra-text css-iziug0"
                >
                  Psycho
                </p>
                <span />
              </div>
              <p
                class="chakra-text css-1lk7enj"
              >
                ‘The ReVe Festival’ Finale
              </p>
              <a
                href="/artists/1z4g3DjTBBZKhvAroFlhOM"
              >
                <p
                  class="chakra-text css-lwat8c"
                >
                  Red Velvet
                </p>
              </a>
            </div>
          </div>
          <div
            class="css-13ce5j0"
          >
            <p
              class="chakra-text css-19ldtcn"
            >
              3:31
            </p>
            <i
              class="bi bi-play-fill song-card-icon"
              data-testid="play-icon"
              id="3CYH422oy1cZNoo0GTG1TK"
              role="play-button"
            />
          </div>
        </div>
      </div>
      <div
        data-testid="song-container"
      >
        <div
          class="css-16l185a"
        >
          <div
            class="css-k008qs"
          >
            <img
              class="chakra-image css-1oaqw3l"
              loading="lazy"
              src="https://i.scdn.co/image/ab67616d000048513f30a062dafcdbc1a8fad842"
            />
            <div
              class="css-13yz2if"
            >
              <div
                class="css-0"
              >
                <p
                  class="chakra-text css-iziug0"
                >
                  러시안 룰렛 Russian Roulette
                </p>
                <span />
              </div>
              <p
                class="chakra-text css-1lk7enj"
              >
                Russian Roulette - The 3rd Mini Album
              </p>
              <a
                href="/artists/1z4g3DjTBBZKhvAroFlhOM"
              >
                <p
                  class="chakra-text css-lwat8c"
                >
                  Red Velvet
                </p>
              </a>
            </div>
          </div>
          <div
            class="css-13ce5j0"
          >
            <p
              class="chakra-text css-19ldtcn"
            >
              3:31
            </p>
            <i
              class="bi bi-play-fill song-card-icon"
              data-testid="play-icon"
              id="5HiSc2ZCGn8L3cH3qSwzBT"
              role="play-button"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="song-container"
    >
      <div
        class="css-16l185a"
      >
        <div
          class="css-k008qs"
        >
          <img
            class="chakra-image css-1oaqw3l"
            loading="lazy"
            src="https://i.scdn.co/image/ab67616d00004851df5022bdf1ac4bf52135c4be"
          />
          <div
            class="css-13yz2if"
          >
            <div
              class="css-0"
            >
              <p
                class="chakra-text css-iziug0"
              >
                Psycho
              </p>
              <span />
            </div>
            <p
              class="chakra-text css-1lk7enj"
            >
              ‘The ReVe Festival’ Finale
            </p>
            <a
              href="/artists/1z4g3DjTBBZKhvAroFlhOM"
            >
              <p
                class="chakra-text css-lwat8c"
              >
                Red Velvet
              </p>
            </a>
          </div>
        </div>
        <div
          class="css-13ce5j0"
        >
          <p
            class="chakra-text css-19ldtcn"
          >
            3:31
          </p>
          <i
            class="bi bi-play-fill song-card-icon"
            data-testid="play-icon"
            id="3CYH422oy1cZNoo0GTG1TK"
            role="play-button"
          />
        </div>
      </div>
    </div>
    <div
      data-testid="song-container"
    >
      <div
        class="css-16l185a"
      >
        <div
          class="css-k008qs"
        >
          <img
            class="chakra-image css-1oaqw3l"
            loading="lazy"
            src="https://i.scdn.co/image/ab67616d000048513f30a062dafcdbc1a8fad842"
          />
          <div
            class="css-13yz2if"
          >
            <div
              class="css-0"
            >
              <p
                class="chakra-text css-iziug0"
              >
                러시안 룰렛 Russian Roulette
              </p>
              <span />
            </div>
            <p
              class="chakra-text css-1lk7enj"
            >
              Russian Roulette - The 3rd Mini Album
            </p>
            <a
              href="/artists/1z4g3DjTBBZKhvAroFlhOM"
            >
              <p
                class="chakra-text css-lwat8c"
              >
                Red Velvet
              </p>
            </a>
          </div>
        </div>
        <div
          class="css-13ce5j0"
        >
          <p
            class="chakra-text css-19ldtcn"
          >
            3:31
          </p>
          <i
            class="bi bi-play-fill song-card-icon"
            data-testid="play-icon"
            id="5HiSc2ZCGn8L3cH3qSwzBT"
            role="play-button"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
